### PR TITLE
feat: Migrate Product Pages to Inertia

### PR DESCRIPTION
**Pull request by bruhhgnik**
Addresses #3061

Migrate the links#show action to render Inertia pages instead of using react_component
helper. Different Inertia pages are rendered based on the layout parameter:

- Products/Show - default product page
- Products/Profile/Show - when layout=profile
- Products/Discover/Show - when layout=discover
- Products/Iframe/Show - when embed or overlay params are present

## Problem

The product page (links#show) uses the legacy react_component helper for rendering, which
is inconsistent with the newer Inertia.js pattern being adopted across the codebase.

## Solution

Migrated the links#show action to render Inertia pages:

**Controller changes:**

- Updated LinksController#show to use `render inertia:` instead of view template
- Added `product_inertia_template` helper to select the correct Inertia page based on params
- Moved props building logic to `ProductPresenter#inertia_page_props` method
- Removed `before_action :hide_layouts` for show action (per @Pradumn27's feedback)

**Frontend changes:**

- Created 4 new Inertia page components under `pages/Products/` directory (consistent with existing Products/Index, Products/New)
- Created shared `ProductPageHead` component that exports reusable `ProductPageHead`, `ProductPageNoScript`, and `ProductPageAlert` - all 4 page variants use this to avoid code duplication
- Product pages use `loggedInUserLayout = true` to get `LoggedInUserProvider` and `CurrentSellerProvider` contexts (per @EmCousin's feedback to use separate layout instead of modifying app_wrapper.tsx)
- SEO meta tags (canonical, structured_data, title) are handled server-side via `set_meta_tag` and `set_product_page_meta` (per @EmCousin's feedback)
- Removed legacy `app/views/links/show.html.erb` and old React on Rails server components

**Screenshots**
Screenshot from 2026-01-30 23-37-37
Screenshot from 2026-01-30 23-37-47
Screenshot from 2026-01-30 23-41-37
Screenshot from 2026-01-30 23-41-53
Screenshot from 2026-01-30 23-42-52

**Tests:**

- Added explicit tests for each layout variant to verify correct Inertia component is rendered
- Profile layout test also verifies `creator_profile` prop is present

## Reviewer Feedback Addressed

This PR addresses feedback from the previous PR #3094:

- **@Pradumn27**: Removed `before_action :hide_layouts` entirely since only show action used it
- **@EmCousin**: 
  - Removed canonical, structured_data, title from client-side props (already handled server-side in `prepare_product_page`)
  - Moved `product_inertia_page_props` logic to `ProductPresenter#inertia_page_props` to relieve the controller
  - Reverted app_wrapper.tsx changes - using `loggedInUserLayout = true` on pages instead

## Test Results

Screenshot from 2026-01-30 23-15-41

## Checklist

- [x] I have read the https://github.com/antiwork/gumroad/blob/main/CONTRIBUTING.md
- [x] I have watched https://www.youtube.com/@anti-work
- [x] I have performed a self-review and left review comments on my PR
- [x] I have added/updated tests for my changes
- [x] I have added screenshots

## AI Disclosure

**Model:** Claude 3.5 Sonnet

**Used for:**
- Generating boilerplate test cases for Inertia template rendering
- Debugging a type error with structured_data prop (backend returning {} instead of array)
- Suggesting the Array.isArray() check for handling edge cases
- Identifying that LoggedInUserProvider was needed for pages with disableLayout

**What I did myself:**
- Designed the component structure and decided on pages/Products/ directory naming
- Wrote all 4 Inertia page components and the shared ProductPageHead
- Controller logic including helper methods for template selection and props building
- Structured props with nested meta object for cleaner separation of concerns
- Reviewed and verified all layout variants locally
- Performed self-review and added comments explaining key implementation decisions

### CODE REVIEW

**Comment by bruhhgnik** [comment]
Hey @EmCousin i fixed all the previously stated issues and here's the new PR, good to merge

**Comment by bruhhgnik** [comment]
## Key Changes Addressing Previous Review Feedback

### 1. SEO Meta Tags Server-Side
All SEO meta tags (`canonical`, `structured_data`, `title`) are now handled entirely server-side via `set_meta_tag` and `set_product_page_meta` callbacks. The client only receives `custom_styles` for seller profile styling.

### 2. No `before_action :hide_layouts`
Removed the `@hide_layouts = false` workaround. The `show` action now simply uses `layout "inertia"` like other Inertia actions.

### 3. Products Namespace
All Inertia pages use the `Products/` namespace (`Products/Show`, `Products/Profile/Show`, etc.) following the codebase convention of renaming `links` to `products`.

### 4. `loggedInUserLayout` Pattern
Instead of modifying the shared `app_wrapper.tsx`, each page exports `loggedInUserLayout = true` to get the `LoggedInUserProvider` and `CurrentSellerProvider` context wrappers.

### 5. Props Logic in Presenter
The `inertia_page_props` method was moved to `ProductPresenter` to keep the controller thin and follow existing patterns.

File: app/javascript/pages/Products/Show.tsx [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+      <ProductPageHead meta={meta} />
+      <ProductPageNoScript />
+      <ProductPageAlert />
+      <Layout {...product} />
+      <PoweredByFooter />
+    </>
+  );
+};
+
+ProductShowPage.loggedInUserLayout = true;
```
Comment (bruhhgnik) [inline app/javascript/pages/Products/Show.tsx:34]: Using `loggedInUserLayout = true` here instead of modifying the shared app_wrapper.tsx. This follows the existing pattern used by other Inertia pages like DownloadPage.

File: app/presenters/product_presenter.rb [hunk 2]
```diff
@@ -65,6 +65,21 @@ def product_page_props(seller_custom_domain_url:, **kwargs)
     }
   end
 
+  def inertia_page_props(layout:, is_embed: false, discover_props: nil, seller_custom_domain_url:, **kwargs)
```
Comment (bruhhgnik) [inline app/presenters/product_presenter.rb:68]: Moved the props assembly logic here from the controller to keep LinksController thin. This method handles the layout-specific props merging (profile layout gets creator_profile, discover gets taxonomy data, etc.).

File: app/javascript/pages/Products/ProductPageHead.tsx [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+import * as React from "react";
+
+import ToastAlert from "$app/components/server-components/Alert";
+
+export type ProductPageMeta = {
+  custom_styles: string | null;
+};
+
+type ProductPageHeadProps = {
+  meta: ProductPageMeta;
```
Comment (bruhhgnik) [inline app/javascript/pages/Products/ProductPageHead.tsx:11]: The meta prop now only contains `custom_styles` - all SEO tags (canonical, structured_data, title) are handled server-side via `set_meta_tag` in the controller.

File: app/controllers/links_controller.rb [hunk 4]
```diff
@@ -33,7 +32,7 @@ class LinksController < ApplicationController
   before_action :fetch_product_and_enforce_ownership, only: %i[destroy]
   before_action :fetch_product_and_enforce_access, only: %i[update publish unpublish release_preorder update_sections]
```
Comment (bruhhgnik) [inline app/controllers/links_controller.rb:33]: Added `show` to the inertia layout list. No need for `before_action :hide_layouts` workaround anymore.

File: app/controllers/links_controller.rb [hunk 5]
```diff
@@ -161,8 +159,9 @@ def show
     end
 
     set_noindex_header if !@product.alive?
+
     respond_to do |format|
-      format.html
+      format.html { render inertia: product_inertia_template, props: product_inertia_props }
```
Comment (bruhhgnik) [inline app/controllers/links_controller.rb:164]: Clean separation: `product_inertia_template` determines which Inertia page to render based on layout/embed params, while `product_inertia_props` calls the presenter for props assembly.

File: app/controllers/links_controller.rb [hunk 6]
```diff
@@ -568,6 +567,28 @@ def prepare_product_page
       @debug                 = params[:debug] && !Rails.env.production?
     end
 
+    def product_inertia_template
+      if params[:layout] == Product::Layout::PROFILE
+        "Products/Profile/Show"
+      elsif params[:layout] == Product::Layout::DISCOVER
```
Comment (bruhhgnik) [inline app/controllers/links_controller.rb:573]: Following the codebase convention of renaming links → products. All templates use the Products/ namespace.

File: app/presenters/product_presenter.rb [hunk 7]
```diff
@@ -65,6 +65,21 @@ def product_page_props(seller_custom_domain_url:, **kwargs)
...
+    when Product::Layout::PROFILE
+      base_props.merge(
+        creator_profile: ProfilePresenter.new(pundit_user:, seller: product.user).creator_profile
+      )
+    when Product::Layout::DISCOVER
+      base_props.merge(discover_props || {})
+    else
+      base_props
+    end
+  end
```
Comment (bruhhgnik) [inline app/presenters/product_presenter.rb:81]: Layout-specific props assembly moved here from the controller. Profile layout merges creator_profile, Discover merges taxonomy props, etc. Keeps the controller thin and follows the existing presenter pattern.

File: app/javascript/pages/Products/ProductPageHead.tsx [hunk 8]
```diff
@@ -0,0 +1,35 @@
+import { Head } from "@inertiajs/react";
+import * as React from "react";
+
+import ToastAlert from "$app/components/server-components/Alert";
+
+export type ProductPageMeta = {
+  custom_styles: string | null;
```
Comment (bruhhgnik) [inline app/javascript/pages/Products/ProductPageHead.tsx:7]: Meta only contains custom_styles now. SEO tags (canonical, structured_data, title) are handled server-side via set_meta_tag and set_product_page_meta callbacks - no client-side duplication needed.

**Comment by bruhhgnik** [comment]
@EmCousin , Fixed all the issues - good to merge now!

**Comment by EmCousin** [comment]
@bruhhgnik ref. https://github.com/antiwork/gumroad/pull/3267/changes
I think you should not be keeping Profile/WishlistPage, correct?

**Comment by bruhhgnik** [comment]
> @bruhhgnik ref. https://github.com/antiwork/gumroad/pull/3267/changes I think you should not be keeping Profile/WishlistPage, correct?

ah i got a bit confused, let me undo that

**Comment by bruhhgnik** [comment]
> > @bruhhgnik ref. https://github.com/antiwork/gumroad/pull/3267/changes I think you should not be keeping Profile/WishlistPage, correct?
> 
> ah i got a bit confused, let me undo that

fixed it

**Comment by bruhhgnik** [comment]
@EmCousin Sir, could you kindly review this PR

**Comment by bruhhgnik** [comment]
Hey @EmCousin can you please review this PR?

CC @Pradumn27

**Comment by bruhhgnik** [comment]
Hey @EmCousin @Pradumn27 can you please review this PR

File: app/controllers/links_controller.rb (hunk 5 above)
Comment (EmCousin) [reply app/controllers/links_controller.rb:164]: It's too much coupling. I'd rather have a proper set of if statements depending on the layout and render the correct template with a dedicated props method for each template.
```ruby
case params[:layout]
when Product::Layout::PROFILE
  render inertia: "Products/Profile/Show", props: presenter.profile_product_props
when Product::Layout::DISCOVER
  render inertia: "Products/Discover/Show", props: presenter.discover_product_props
# [...] etc.
end
```

File: app/javascript/pages/Products/ProductPageHead.tsx [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+  const customStyles = meta.custom_styles ?? "";
+
+  if (customStyles.trim().length === 0) return null;
+
+  return (
+    <Head>
+      <style dangerouslySetInnerHTML={{ __html: customStyles }} />
+    </Head>
+  );
+};
```
Comment (EmCousin) [inline app/javascript/pages/Products/ProductPageHead.tsx:24]: Since those styles are inserted in the `<head>`, can you try to render them server-side via the `set_meta_tag` helper in the controller? Something like this
```ruby
set_meta_tag(tag_name: "style", inner_content: product.user.seller_profile.custom_styles.to_s)
```

This will avoid layout shifts since the styles would otherwise be added after the initial load page.

File: app/javascript/packs/product.ts [hunk 10]
```diff
@@ -2,17 +2,9 @@ import ReactOnRails from "react-on-rails";
 
 import BasePage from "$app/utils/base_page";
 
-import DiscoverProductPage from "$app/components/server-components/Discover/ProductPage";
-import ProductPage from "$app/components/server-components/Product";
-import ProductIframePage from "$app/components/server-components/Product/IframePage";
 import ProfileCoffeePage from "$app/components/server-components/Profile/CoffeePage";
-import ProfileProductPage from "$app/components/server-components/Profile/ProductPage";
```
Comment (EmCousin) [inline app/javascript/packs/product.ts:9]: You forgot to delete that file

File: app/presenters/product_presenter.rb (hunk 7 above)
Comment (EmCousin) [reply app/presenters/product_presenter.rb:81]: This has not been tested
